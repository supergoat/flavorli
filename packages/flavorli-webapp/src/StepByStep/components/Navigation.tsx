import React from 'react';
import {Button, Stack} from '@flavorli/elements';
import styled, {FlattenSimpleInterpolation} from 'styled-components';
import ChevronRightWhite from '../icons/right_chevron_white.svg';

interface INavigation {
  onNavigate: (direction: 1 | -1) => void;
  hideNextStepButton?: boolean;
  hideBackButton?: boolean;
  nextButtonName?: string;
  backButtonName?: string;
  variation?: 'onPrimary';
}
const Navigation = ({
  onNavigate,
  hideNextStepButton = false,
  hideBackButton = false,
  nextButtonName,
  backButtonName,
  variation,
}: INavigation) => {
  const navigateToNextStep = () => {
    onNavigate(1);
  };

  const navigateToPreviousStep = () => {
    onNavigate(-1);
  };

  return (
    <Stack direction="horizontal" gap={8} distribution="end" width="100%">
      <PreviousButton
        hide={hideBackButton}
        aria-controls="recipe-steps"
        tabIndex={hideBackButton ? -1 : undefined}
        onClick={navigateToPreviousStep}
        intent={variation === 'onPrimary' ? 'secondaryOnPrimary' : 'secondary'}
      >
        {backButtonName || 'back'}
      </PreviousButton>

      {!hideNextStepButton && (
        <Button
          aria-controls="recipe-steps"
          width="100%"
          onClick={navigateToNextStep}
          intent={variation === 'onPrimary' ? 'secondaryOnPrimary' : 'primary'}
        >
          {nextButtonName || 'Next'}
          <ButtonIcon />
        </Button>
      )}
    </Stack>
  );
};

export default Navigation;

const PreviousButton = styled(Button)<{
  hide: boolean;
  css?: FlattenSimpleInterpolation;
}>`
  ${p => p.css && p.css};
  visibility: ${p => (p.hide ? 'hidden' : 'visible')};
`;

const ButtonIcon = styled.img.attrs(() => ({
  src: ChevronRightWhite,
  alt: '',
}))`
  position: absolute;
  right: 15px;
`;
