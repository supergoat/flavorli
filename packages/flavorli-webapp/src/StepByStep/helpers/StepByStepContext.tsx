import React from 'react';

interface IStepsContext {
  currentDialogStep: number;
  currentStep: number;
  noOfSteps: number;
  onOpenDialogStep: (stepNo: number) => void;
  onCloseDialogStep: () => void;
  onNavigate: (direction: 1 | -1) => void;
}

export const StepByStepContext = React.createContext<IStepsContext | null>(
  null,
);

export function StepByStepProvider({
  initialValues,
  ...props
}: {
  initialValues: {
    noOfSteps: number;
    currentStep?: number;
  };
  children: React.ReactNode;
}) {
  const [currentStep, setCurrentStep] = React.useState(
    initialValues?.currentStep || 1,
  );
  const [currentDialogStep, setCurrentDialogStep] = React.useState();
  const [lastFocus, setLastFocus] = React.useState();

  const onOpenDialogStep = (stepNo: number) => {
    setLastFocus(document.activeElement);
    setCurrentDialogStep(stepNo);
  };

  const onCloseDialogStep = () => {
    setCurrentDialogStep(null);
    lastFocus.focus();
  };

  const onNavigate = (direction: 1 | -1) => {
    setCurrentStep(s => s + direction);
  };

  return (
    <StepByStepContext.Provider
      value={{
        currentStep,
        currentDialogStep,
        onOpenDialogStep,
        onCloseDialogStep,
        onNavigate,
        noOfSteps: initialValues.noOfSteps,
      }}
      {...props}
    />
  );
}

export function useStepsContext() {
  const context = React.useContext(StepByStepContext);

  if (!context) {
    throw new Error('useStepsContext must be used within a StepsProvider');
  }
  return context;
}
