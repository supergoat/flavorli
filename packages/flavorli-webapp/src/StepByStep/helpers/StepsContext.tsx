import React from 'react';

interface IStepsContext {
  openLink: number;
  currentStep: number;
  noOfSteps: number;
  onViewStep: (stepNo: number) => void;
  onCloseViewStep: () => void;
  onChangeStep: (direction: 1 | -1) => void;
}

export const StepByStepContext = React.createContext<IStepsContext | null>(
  null,
);

export function StepByStepProvider({
  initialValues,
  ...props
}: {
  initialValues: {
    noOfSteps: number;
    currentStep?: number;
  };
  children: React.ReactNode;
}) {
  const [currentStep, setCurrentStep] = React.useState(
    initialValues?.currentStep || 1,
  );
  const [openLink, setOpenLink] = React.useState();
  const [lastFocus, setLastFocus] = React.useState();

  const onViewStep = (stepNo: number) => {
    setLastFocus(document.activeElement);
    setOpenLink(stepNo);
  };

  const onCloseViewStep = () => {
    setOpenLink(null);
    lastFocus.focus();
  };

  const onChangeStep = (direction: 1 | -1) => {
    setCurrentStep(s => s + direction);
  };

  return (
    <StepByStepContext.Provider
      value={{
        currentStep,
        openLink,
        onViewStep,
        onCloseViewStep,
        onChangeStep,
        noOfSteps: initialValues.noOfSteps,
      }}
      {...props}
    />
  );
}

export function useStepsContext() {
  const context = React.useContext(StepByStepContext);

  if (!context) {
    throw new Error('useStepsContext must be used within a StepsProvider');
  }
  return context;
}
