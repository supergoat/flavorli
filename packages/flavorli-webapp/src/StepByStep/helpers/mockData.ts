import {IStep} from '../types';

export const steps: IStep[] = [
  {
    no: 1,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [],
    description: 'Place a pot of water in medium heat and set aside to boil',
    ingredients: [],
    kitchenware: [
      {
        qty: '1',
        name: 'Pot',
      },
    ],
  },
  {
    no: 2,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [],
    description: 'Place a sauce pan over high heat and add the olive oil',
    ingredients: [
      {
        qty: '4 - 6 tbsps',
        name: 'olive oil',
      },
    ],
    kitchenware: [
      {
        qty: '1',
        name: 'Sauce Pan',
      },
    ],
  },
  {
    no: 3,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [],
    description:
      'Cut the eggplan into 1 cm cubes. Add it to the pan and sauté for 1-2 minutes',
    ingredients: [
      {
        qty: '2',
        name: 'eggplants',
      },
    ],
    kitchenware: [
      {qty: '1', name: 'Knife'},
      {qty: '1', name: 'Cutting Board'},
      {qty: '1', name: 'Spoon'},
    ],
  },
  {
    no: 4,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [],
    description:
      'Finely chop the onion, add it to the pan and sauté for 1-2 minutes',
    ingredients: [
      {
        qty: '1',
        name: 'onion',
      },
    ],
    kitchenware: [],
  },
  {
    no: 5,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [],
    description:
      'Cut the zucchinis and the peppers into small cubes and add them to the pan. Sauté for 4-5 minutes, until the vegetables are ready',
    ingredients: [
      {
        qty: '2',
        name: 'zucchinis',
      },
      {
        qty: '1',
        name: 'red bell pepper',
      },
      {
        qty: '1',
        name: 'green bell pepper',
      },
    ],
    kitchenware: [],
  },
  {
    no: 6,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [],
    description:
      'Add the salt, pepper, oregano, thyme, chili flakes, sugar and tomato paste and sauté',
    ingredients: [
      {
        qty: '',
        name: 'salt',
      },
      {
        qty: '',
        name: 'pepper',
      },
      {
        qty: '1 tsp',
        name: 'oregano',
      },
      {
        qty: '1 tbsp',
        name: 'thyme',
      },
      {
        qty: '1 tsp',
        name: 'chilli flakes',
      },
      {
        qty: '1 tbsp',
        name: 'granulated sugar',
      },
      {
        qty: '1 tsp',
        name: 'tomato paste',
      },
    ],
    kitchenware: [
      {
        qty: '1',
        name: 'Tea Spoon',
      },
    ],
  },
  {
    no: 7,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [],
    description:
      'Add the wine and canned tomatoes and simmer at medium-low heat for 10 minutes',
    ingredients: [
      {
        qty: '40 g',
        name: 'white wine',
      },
      {
        qty: '400 g',
        name: 'canned tomatoes',
      },
    ],
    timer: {id: 7, name: 'Step 7: Simmer Sauce', minutes: 10, seconds: 0},
    kitchenware: [{qty: '1', name: 'scale'}],
  },
  {
    no: 8,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [{from: 1, heading: 'From Step 1', name: 'Boiling water'}],
    description:
      'In the pot with boiling water, add salt along with the macaroni, and boil for ten minutes',
    ingredients: [
      {
        qty: '',
        name: 'salt',
      },
      {
        qty: '500 g',
        name: 'macaroni',
      },
    ],
    timer: {id: 8, name: 'Step 8: Cook Pasta', minutes: 10, seconds: 0},
    kitchenware: [],
  },
  {
    no: 9,
    tag: {
      text: 'Assembling',
      color: 'tagPink',
    },
    links: [],
    description: 'Preheat the oven at 180 C set to fan',
    ingredients: [],
    kitchenware: [],
  },
  {
    no: 10,
    tag: {
      text: 'Bechamel',
      color: 'tagOrange',
    },
    links: [],
    description:
      'Place a pot over medium heat. Add the olive oil and the flour and using a hand whisk, mix well until all of the flour is absorbed',
    ingredients: [
      {
        qty: '80 g',
        name: 'olive oil',
      },
      {
        qty: '80 g',
        name: 'all-purpose flour',
      },
    ],
    kitchenware: [
      {qty: '1', name: 'Small Pot'},
      {qty: '1', name: 'Hand Whisk'},
    ],
  },
  {
    no: 11,
    tag: {
      text: 'Bechamel',
      color: 'tagOrange',
    },
    links: [],
    description: 'Add the soy milk in 7-8 batches, stirring constantly',
    ingredients: [
      {
        qty: '800 g',
        name: 'soy milk',
      },
    ],
    kitchenware: [],
  },
  {
    no: 12,
    tag: {
      text: 'Bechamel',
      color: 'tagOrange',
    },
    links: [],
    description:
      'Season with salt and pepper and as soon as it thickens and comes to boil, remove from the heat',
    ingredients: [
      {
        qty: '',
        name: 'salt',
      },
      {
        qty: '',
        name: 'pepper',
      },
    ],
    kitchenware: [],
  },
  {
    no: 13,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [
      {
        from: 7,
        heading: 'From Step 7',
        name: 'Pan with vegetables',
        timerId: 7,
      },
      {
        from: 8,
        heading: 'From Step 8',
        name: 'Macaroni',
        timerId: 8,
      },
    ],
    description:
      'Drain the macaroni and transfer to the pan with the vegetables. Mix and set aside',
    ingredients: [],
    kitchenware: [{qty: '1', name: 'Colander'}],
  },
  {
    no: 14,
    tag: {
      text: 'Assembling',
      color: 'tagPink',
    },
    links: [{from: 12, heading: 'From Step 12', name: 'Bechamel sauce'}],
    description:
      'Add 100g of the bechamel into the pasta mixture, add oregano, mint, and mix well',
    ingredients: [
      {
        qty: '1 tsp',
        name: 'oregano',
      },
      {
        qty: '1 tbsp',
        name: 'mint',
      },
    ],
    kitchenware: [],
  },
  {
    no: 15,
    tag: {
      text: 'Assembling',
      color: 'tagPink',
    },
    links: [],
    description:
      'Grease 40 cm ovenproof baking dish with olive oil and sprinke the breadcrumps',
    ingredients: [
      {
        qty: '1 tbsp',
        name: 'olive oil',
      },
      {
        qty: '25 g',
        name: 'breadcrumps',
      },
    ],
    kitchenware: [{qty: '1', name: '40 cm ovenproof baking dish'}],
  },
  {
    no: 16,
    tag: {
      text: 'Assembling',
      color: 'tagPink',
    },
    links: [],
    description:
      'Add the pasta mixture, spread the bechamel sauce, and sprinke the breadcrumps, olive oil, thyme and pepper',
    ingredients: [
      {
        qty: '25 g',
        name: 'breadcrumps',
      },
      {
        qty: '1 tbsp',
        name: 'olive oil',
      },
      {
        qty: '1 tbsp',
        name: 'thyme',
      },
      {
        qty: '',
        name: 'pepper',
      },
    ],
    kitchenware: [],
  },
  {
    no: 17,
    tag: {
      text: 'Assembling',
      color: 'tagPink',
    },
    links: [],
    description: 'Bake for 25 - 30 minutes. Remove and let it cool',
    ingredients: [],
    timer: {id: 17, name: 'Step 17: Bake Pastitsio', minutes: 30, seconds: 0},
    kitchenware: [
      {
        qty: '1',
        name: 'Oven Mitts',
      },
    ],
  },
];
