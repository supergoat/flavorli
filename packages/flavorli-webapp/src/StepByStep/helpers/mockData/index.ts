// https://www.youtube.com/watch?v=q1q3HfVvzQ0
export {introStep} from './introStep';
export {ingredientsStep} from './ingredientsStep';
export {itemsStep} from './itemsStep';
export {preparationSteps} from './preparationSteps';
export {recipeSteps} from './recipeSteps';
