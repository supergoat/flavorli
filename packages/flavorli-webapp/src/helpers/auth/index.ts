import {CognitoUserPool} from 'amazon-cognito-identity-js';
import * as AWS from 'aws-sdk/global';
import {authenticate} from './authenticate';
import {createNewUser} from './createNewUser';
import {establishUserSession} from './establishUserSession';
import {verifyPhoneNumber} from './verifyPhoneNumber';

AWS.config.region = 'eu-west-2';

export const userPool = new CognitoUserPool({
  UserPoolId: process.env.REACT_APP_COGNITO_USER_POOL_ID as string,
  ClientId: process.env.REACT_APP_COGNITO_CLIENT_ID as string,
});

export const loginId = `cognito-idp.${
  AWS.config.region
}.amazonaws.com/${userPool.getUserPoolId()}`;

export {authenticate, createNewUser, establishUserSession, verifyPhoneNumber};
