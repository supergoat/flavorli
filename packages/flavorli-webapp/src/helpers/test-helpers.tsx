import React from 'react';
import {Router} from 'react-router-dom';
import {createMemoryHistory} from 'history';
import {ThemeProvider} from 'styled-components';
import {render as testRender} from '@testing-library/react';
import {theme} from '@flavorli/elements';

export function render(ui: React.ReactNode) {
  const utils = testRender(<ThemeProvider theme={theme}>{ui}</ThemeProvider>);

  return utils;
}

export function renderWithRouter(ui: React.ReactElement, route: string = '/') {
  const history = createMemoryHistory({initialEntries: [route]});
  const Wrapper = ({children}: {children: React.ReactNode}) => (
    <Router history={history}>{children}</Router>
  );
  return {
    ...testRender(<ThemeProvider theme={theme}>{ui}</ThemeProvider>, {
      wrapper: Wrapper as React.ComponentType,
    }),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  };
}
