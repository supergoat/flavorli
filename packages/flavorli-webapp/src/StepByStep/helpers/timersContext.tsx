import React from 'react';
import {ITimer} from '../types';
import {useInterval} from '../../helpers/hooks';

interface ITimersContext {
  timers: {[timerId: number]: ITimer};
  setTimers: React.Dispatch<React.SetStateAction<{}>>;
}

export const TimersContext = React.createContext<ITimersContext | null>(null);

export function TimersProvider({
  initialValues,
  ...props
}: {
  initialValues?: Partial<ITimersContext>;
  children: React.ReactNode;
}) {
  const [contextTimers, setTimers] = React.useState({});

  const timers = initialValues?.timers;

  React.useEffect(() => {
    if (timers) setTimers(timers);
  }, [timers]);

  if (!contextTimers) return null;

  return (
    <TimersContext.Provider
      value={{timers: contextTimers, setTimers}}
      {...props}
    />
  );
}

export function useTimersContext() {
  const context = React.useContext(TimersContext);
  if (!context) {
    throw new Error('useTimersContext must be used within a TimersProvider');
  }
  return context;
}

export function useAddTimerIfItDoesNotExist(timer: ITimer) {
  const {timers, setTimers} = useTimersContext();
  if (timers[timer.id]) return;
  setTimers(t => ({...t, [timer.id]: {...timer, isPaused: true}}));
}

export function useUpdateTimers(timer: ITimer) {
  const {setTimers} = useTimersContext();

  useInterval(() => {
    if (timer.isPaused) return;

    if (timer.minutes === 0 && timer.seconds === 0) {
      setTimers(t => ({
        ...t,
        [timer.id]: {...timer, isPaused: true},
      }));
    } else if (timer.seconds > 0) {
      setTimers(t => ({
        ...t,
        [timer.id]: {...timer, seconds: timer.seconds - 1},
      }));
    } else {
      setTimers(t => ({
        ...t,
        [timer.id]: {...timer, minutes: timer.minutes - 1, seconds: 59},
      }));
    }
  }, 1000);
}
