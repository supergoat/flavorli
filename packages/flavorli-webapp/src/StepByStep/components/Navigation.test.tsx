import React from 'react';
import {axe} from 'jest-axe';
import {render} from '../../helpers/test-helpers';
import Navigation from './Navigation';
import userEvent from '@testing-library/user-event';

const setup = ({
  nextButtonName,
  backButtonName,
  hideNextStepButton = false,
  hideBackButton = false,
  varation,
}: {
  nextButtonName?: string;
  backButtonName?: string;
  hideNextStepButton?: boolean;
  hideBackButton?: boolean;
  varation?: 'onPrimary';
} = {}) => {
  const mockOnNavigate = jest.fn();

  return {
    ...render(
      <>
        {/* Add a div with id recipe-steps to be used by aria-controls */}
        <div id="recipe-steps" />
        <Navigation
          nextButtonName={nextButtonName}
          backButtonName={backButtonName}
          onNavigate={mockOnNavigate}
          hideNextStepButton={hideNextStepButton}
          hideBackButton={hideBackButton}
          varation={varation}
        />
      </>,
    ),
    mockOnNavigate,
  };
};

describe('Navigation', () => {
  it('should not have any axe violations', async () => {
    const {container} = setup();
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should render correctly', () => {
    const {container} = setup();
    expect(container).toMatchSnapshot();
  });

  it('should have a next button with aria-controls set to "recipe-steps" to inform assistive technology users that it controls which step is displayed', () => {
    const {getByText} = setup();

    const nextStepButton = getByText(/Next/i);

    expect(nextStepButton).toHaveAttribute('aria-controls', 'recipe-steps');
  });

  it('should accept a custom name for the next button', () => {
    const {getByText} = setup({nextButtonName: 'Continue'});

    getByText(/Continue/i);
  });

  it('clicking the nextStep button calls onNavigate with 1', () => {
    const {getByText, mockOnNavigate} = setup();

    const nextStepButton = getByText(/Next/i);

    userEvent.click(nextStepButton);

    expect(mockOnNavigate).toHaveBeenCalledWith(1);
  });

  it('it should hide the nextStep button when hideNextStepButton is set to true', () => {
    const {queryByText} = setup({hideNextStepButton: true});
    expect(queryByText('Next')).toBeNull();
  });

  it('should have a back button with aria-controls set to "recipe-steps" to inform assistive technology users that it controls which step is displayed', () => {
    const {getByText} = setup();

    const backButton = getByText(/back/i);

    expect(backButton).toHaveAttribute('aria-controls', 'recipe-steps');
  });

  it('should accept a custom name for the back button', () => {
    const {getByText} = setup({backButtonName: 'previous'});

    getByText(/previous/i);
  });

  it('clicking the back button calls onNavigate with -1', () => {
    const {getByText, mockOnNavigate} = setup();

    const backButton = getByText(/back/i);

    userEvent.click(backButton);

    expect(mockOnNavigate).toHaveBeenCalledWith(-1);
  });

  it('it should hide the back button when hideBackButton is set to true', () => {
    const {getByText} = setup({hideBackButton: true});
    expect(getByText(/back/i)).not.toBeVisible();
    expect(getByText(/back/i)).toHaveAttribute('tabIndex', '-1');
  });

  it('should render correctly when variation is onPrimary', () => {
    const {container} = setup({
      varation: 'onPrimary',
    });

    expect(container).toMatchSnapshot();
  });
});
