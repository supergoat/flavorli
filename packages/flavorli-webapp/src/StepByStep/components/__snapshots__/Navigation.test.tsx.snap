// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation should render correctly 1`] = `
<div>
  <div
    id="recipe-steps"
  />
  <div
    class="sc-AykKH hoxMSq sc-fzXfLO fZNNwI"
    direction="horizontal"
    width="100%"
  >
    <button
      aria-controls="recipe-steps"
      class="sc-AykKC sc-fzXfLP cWHKgM"
    >
      back
    </button>
    <button
      aria-controls="recipe-steps"
      class="sc-AykKC eAmRjU"
      width="100%"
    >
      Next
      <img
        alt=""
        class="sc-fzXfLQ jNZvTh"
        src="right_chevron_white.svg"
      />
    </button>
  </div>
</div>
`;

exports[`Navigation should render correctly when variation is onPrimary 1`] = `
<div>
  <div
    id="recipe-steps"
  />
  <div
    class="sc-AykKH hoxMSq sc-fzXfLO fZNNwI"
    direction="horizontal"
    width="100%"
  >
    <button
      aria-controls="recipe-steps"
      class="sc-AykKC sc-fzXfLP cWHKgM"
    >
      back
    </button>
    <button
      aria-controls="recipe-steps"
      class="sc-AykKC eAmRjU"
      width="100%"
    >
      Next
      <img
        alt=""
        class="sc-fzXfLQ jNZvTh"
        src="right_chevron_white.svg"
      />
    </button>
  </div>
</div>
`;
